RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h

```diff
@@ -174,25 +189,81 @@ __global__ void kernel_fastDuplicateRemover(GPUCACell const *__restrict__ cells,
       continue;
     // if (thisCell.theDoubletId < 0) continue;
 
-    float mc = 10000.f;
-    uint16_t im = 60000;
+    float mc = maxScore;
+    uint16_t im = tkNotFound;
 
+    /* chi2 penalize higher-pt tracks  (try rescale it?)
     auto score = [&](auto it) {
-      return std::abs(tracks->tip(it));  // tip
-      // return tracks->chi2(it);  //chi2
+      return foundNtuplets->size(it) < 4 ? 
+              std::abs(tracks->tip(it)) :  // tip for triplets
+              tracks->chi2(it);            //chi2 for quads
     };
+    */
+
+    auto score = [&](auto it) { return std::abs(tracks->tip(it)); };
+
+    // full crazy combinatorics
+    int ntr = thisCell.tracks().size();
+    for (int i = 0; i < ntr; ++i) {
+      auto it = thisCell.tracks()[i];
+      auto qi = tracks->quality(it);
+      if (qi <= reject)
+        continue;
+      auto opi = tracks->stateAtBS.state(it)(2);
+      auto e2opi = tracks->stateAtBS.covariance(it)(9);
+      auto cti = tracks->stateAtBS.state(it)(3);
+      auto e2cti = tracks->stateAtBS.covariance(it)(12);
+      for (auto j = i + 1; j < ntr; ++j) {
+        auto jt = thisCell.tracks()[j];
+        auto qj = tracks->quality(jt);
+        if (qj <= reject)
+          continue;
+#ifdef GPU_DEBUG
+        if (foundNtuplets->size(it) != foundNtuplets->size(jt))
+          printf(" a mess\n");
+#endif
+        auto opj = tracks->stateAtBS.state(jt)(2);
+        auto ctj = tracks->stateAtBS.state(jt)(3);
+        auto dct = 25.f * (tracks->stateAtBS.covariance(jt)(12) + e2cti);
```

Comment:
could the origin of `25.f` be clarified here and elsewhere, and/or made into a named constant?

Reply:
5 sigma. making into a named caonstant does not change anything

Reply:
please consider adding a comment inline to explain this and further improve the readability of the code.